cvat/apps/dataset_manager/task.py

```diff
@@ -103,18 +110,15 @@ def add_prefetch_info(cls, queryset):
             'segment__task__project__owner',
             'segment__task__project__assignee',
 
-            Prefetch('segment__task__data',
-                queryset=models.Data.objects.select_related('video').prefetch_related(
-                    Prefetch('images', queryset=models.Image.objects.order_by('frame'))
-            )),
+            Prefetch('segment__task__data', queryset=task_data_queryset),
 
             Prefetch('segment__task__label_set', queryset=label_qs),
             Prefetch('segment__task__project__label_set', queryset=label_qs),
         )
 
-    def __init__(self, pk, *, is_prefetched=False, queryset=None):
+    def __init__(self, pk, *, is_prefetched: bool = False, queryset: QuerySet = None, prefetch_images: bool = True):
```

Comment:
I can't say, this conditional logic was added in https://github.com/cvat-ai/cvat/commit/ba74709c40c08d31db59814bd4f9abec51ee3282.

> For instance, why don't we lock the row when updating job annotations directly?

I'm not sure what you mean here. Could you phrase it in more detail? I can guess that the whole update was supposed to be a single request, so that no lock is needed. Maybe there was some deadlock somewhere.

Reply:
set prefetch_images=False by default

Reply:
as far as I understand from https://github.com/cvat-ai/cvat/pull/5160/files#diff-ed7ab63c7c54f5d87f982240b298e7830de8e01da28b819779b44bc601db6f7bR74
`is_prefetched` actually is related to my `prefetch_images` - it was to determine whether images should be prefetched. But somewhere along the way this behaviour was broken.

select_for_update was there earlier, from (at least) https://github.com/cvat-ai/cvat/commit/ae6a4891e3fd874331dda661572aceb09a49fdb7 (in cvat/apps/engine/annotation.py) and was removed for the case when images are to be prefetched

Reply:
Since in most cases jobs are not locked, and `is_prefetched` is only  used in `TaskAnnotation.init_from_db`, I believe `is_prefetched` can be dropped.

Reply:
> Since is_prefetched is only used in TaskAnnotation.init_from_db and in all the other cases lock was removed two years ago and no problems emerged (?), I believe that the lock is not required and is_prefetched can be removed.

I do not agree with that. Maybe such problems were. Imagine the situation when 2 parallel PATCH requests come to update task annotations. If we just remove `select_for_update` then the result will be unexpected (At first glance, the same problem exists now for jobs). Probably we should not lock the database job or task row when updating annotations but use lock per annotations update action.

Reply:
I investigated a bit more, and there only two cases when this lock happens on some changes in db: when a task is added to project and when a dataset is imported as a project. In all the other cases, this lock happens when no changes are applied - e.g. task export. 
But the lock is required on export, according to https://github.com/cvat-ai/cvat/blob/develop/cvat/apps/dataset_manager/task.py#L1087
therefore I am returning the option, but with the other, more descriptive name

Reply:
I guess we actually need to figure out when we don't need to lock. There must be at least the following clients of this class:

- job, task, project import, export
- job, task annotation changes (patch, put), honeypot reroll in a job (can remove annotations on the changed honeypot frames)
- quality computation (~export), backup (~export), autoannotation (~import)

It looks to me that in most, if not all, of the cases we do need to lock data. 
- For exporting, the lock is needed to retrieve the relevant data from the DB. If there are multiple DB tables or if there is a batched load from a single table, we do need a lock. We also need image meta
- For importing, the lock is needed because the current logic is to overwrite the existing data. This logic was asked to be changed many times, however, this how it is supposed to be now. Image meta is needed as well, as we need to know image sizes for some of the formats and we need to match image names from the uploaded annotations
- For annotation changes in general, there is a writing step, for sure. But there is also a reading step, which is needed to compute the diff, and record it for analytics

We don't need to hold the lock for the whole operation though. The lifetime of a lock is bound to the transaction, and they are used to control it in the code. For exporting, we typically need it only to prefetch everything in the beginning. For importing, we probably need to lock 2 times - to read and to (over-)write.